src/api/getAmmFundingOrderData.js

```diff
@@ -15,6 +15,12 @@ const validateData = validateWithJoi(schema)('INVALID_METHOD_ARGUMENT')({
 
 const endpoint = '/v1/trading/amm/fundingOrderData'
 
-module.exports = (dvf, data, nonce, signature) => get(
-  dvf, endpoint, nonce, signature, validateData(data)
-)
+module.exports = (dvf, data, nonce, signature) => {
+  // convert `amount` to string, since it's passed as a query string parameter, and bigNumber vars aren't converted
+  const validatedData = validateData(data)
+  validatedData.amount = validatedData.amount.toString()
```

Comment:
as a general rule is best to avoid object mutation

We are assuming here that `validateData` doesn't return the unchanged `data` object. If it did then we'd be mutating a function argument, which is obviously not a good idea. Even is right now `validateData` always returns a clone of data it could change in the future which could lead to a difficult to diagnose bugs.

Reply:
How are we supposed to convert `amount` to string then?
As far as I understood from the above conversation, the problem was that `bigNumber` values are not converted to string when passed as part of the `query string`. And this does happen when they're part of the `body` in a POST request. So, our goal is to mimic this behaviour.

I'm not sure I understand what you mean by
> We are assuming here that validateData doesn't return the unchanged data object. 

The function _does_ return the validated and parsed `data` object.
```
const { value: validated, error } = schema.validate(value, options)
...
return validated
```

If the return value of `validateData` changes in the future, that would render pretty much all endpoints that use it broken, so it won't be an issue of this particular usage only.

Reply:
>We are assuming here that validateData doesn't return the unchanged data object. If it did then we'd be mutating a function argument,

By that assumption of change in `validateData`, we would need to clone `data` before passing it in, since if `validateData`a mutates it, the additional mutation of `amount` is only secondary and this function is already in violation of mutating the argument through the call to `validateData`. However `validateData` is returning a new cloned object, if it changes in the future, wouldn't that be considered a bug on the validation function since it is expected to not mutate the arguments passed in/
It would be similar to assuming `R.append` could be changed to mutate the array passed in which breaks the contract of the function in the first place

Reply:
In order for us to know whether the validation returns the same `data` object, but not a clone, in case it doesn't parse any of the props, we have to know how `schema.validate()` works - aka the internals of `Joi`.

If we want to safeguard against mutations of function props, as you @adrian-gierakowski described, it would be much better to return a clone from `validateData` instead of having to think about this scenario every time we use that function (aka at the receiving side of that function). So, we could just have the following implementation of `validateData`:
```
  const { value: validated, error } = schema.validate(value, options)
 ...
  return { ...validated }
```
This way we cover this edge case faulty scenario for all callees.

But IMO this whole safeguarding is a bit pointless, since we will _always_ pass to the request function a `data` object that has been validated. 

I don't like the [code snippet in the above comment](https://github.com/DeversiFi/dvf-client-js/pull/272#discussion_r713052440), because it's needlessly complex and reduces code readability.

Anyway, I will create a clone of the validated `data` object thru object destruction, as you suggested, so we avoid ambiguous bugs. 
But if we want to safeguard this _the proper way_, I believe this object cloning should happen in `validateData()`.

Reply:
those snippets focus on composability, hence they might look complex to read if chained in main logic however if scoped, they'll be much easier and cleaner to read

```js
// compose the prop conversion into a named variable/function
const stringifyProp = prop => R.over(R.lensProp(prop), R.toString)
// or stringifyAmount and use 'amount' literal instead of prop in there
const data = { a: 1, b: 2}
const convertedData = stringifyProp('a')(data) // { a: '1', b: 2 }
```

Another point regarding `{ ...validated }`, this is not really cloning the object, it is only a shallow copy and any objects inside will be copied by reference, in order to truely clone, you'll need something like `R.clone`, those operations could be fairly expensive (you can profile it in a script), and best avoided if the object is already cloned

Reply:
> In order for us to know whether the validation returns the same `data` object, but not a clone, in case it doesn't parse any of the props, we have to know how `schema.validate()` works - aka the internals of `Joi`.

Yes, exactly! This is the point I was trying to make. We should write code which does the right thing without having to think about internal implantation details of functions used to produce the inputs or consume outputs of the code in question.

> If we want to safeguard against mutations of function props, as you @adrian-gierakowski described, it would be much better to return a clone from validateData instead of having to think about this scenario every time we use that function 
 
You are missing the point: cloning or not cloning within `validateData` is an internal implementation detail of this function, the same way as in the case of the above discussed `schema.validate()`. Code using `validateData` should not care about it's internal implementation details. Following the non-mutation principle frees you from having to think about implementation details of the surrounding code.

Reply:
> > We are assuming here that validateData doesn't return the unchanged data object. If it did then we'd be mutating a function argument,
> 
> By that assumption of change in `validateData`, we would need to clone `data` before passing it in, since if `validateData`a mutates it, the additional mutation of `amount` is only secondary and this function is already in violation of mutating the argument through the call to `validateData`. However `validateData` is returning a new cloned object, if it changes in the future, wouldn't that be considered a bug on the validation function since it is expected to not mutate the arguments passed in?
> It would be similar to assuming `R.append` could be changed to mutate the array passed in which breaks the contract of the function in the first place

@arijoon looks like I didn't word this well. What I meant to say that the mutating implementation would only be safe if we assume that `validateData` returns a clone of `data`. I didn't mean to say that `validateData` could be refactored to mutate `data`. I meant that if `validateData` in some cases simply passes `data` through (unchanged) then by mutating it's output we'd be mutating the `data` argument of the enclosing function (which I believe we all agree is unsafe in general).

Reply:
> Another point regarding `{ ...validated }`, this is not really cloning the object, it is only a shallow copy

correct, however it's good enough for this purpose as we are only modifying direct child prop of the object

for modification of deeply nested props [R.assocPath](https://ramdajs.com/docs/#assocPath) as it performs the minimum amount of cloning necessary to modify a value at given path without mutating any objects